File: ahc030/a/tools/main.py
import sys
from collections import deque
from itertools import accumulate

input = lambda: sys.stdin.readline().rstrip()
P = lambda *x: print(*x)
PY = lambda: print("Yes")
PN = lambda: print("No")
II = lambda: int(input())
MII = lambda: map(int, input().split())
LMII = lambda: list(map(int, input().split()))
around4 = ((-1, 0), (1, 0), (0, -1), (0, 1))


def dlist(*l, fill=0):
    if len(l) == 1:
        return [fill] * l[0]
    ll = l[1:]
    return [dlist(*ll, fill=fill) for _ in range(l[0])]


# 毎回Dをnextの数でソートする
def FirstInput():
    global N, M, E, D, D_size, limit, kakutei, D_size_sum
    N, M, E = input().split()
    N, M = int(N), int(M)
    E = float(E)
    D = []

    kakutei = [False] * (N**2)
    D_size_sum = 0
    limit = 2100000 // N**2
    for _ in range(M):
        tmp = LMII()
        zahyou = []
        x_max, y_max = 0, 0
        for i in range(2, len(tmp), 2):
            x_max = max(x_max, tmp[i - 1])
            y_max = max(y_max, tmp[i])
            zahyou.append((tmp[i - 1], tmp[i]))
        D.append([zahyou, len(zahyou), [True] * (N**2), [x_max, y_max]])
        D_size_sum += len(zahyou)

    D.sort(key=lambda x: x[1], reverse=True)

    D_size = [i for _, i, _, _ in D]
    D_size = list(accumulate(D_size[::-1]))[::-1]
    # print("#", D_size_sum / 4)


def Output_Input(l: list):
    tmp = []
    for i, j in l:
        tmp.append(i)
        tmp.append(j)
    print("q", len(l), *tmp, flush=True)
    n = II()
    return n


def Output_Input_add(l: list, i, j, num):
    tmp = []
    tmp.append(i)
    tmp.append(j)
    for i, j in l:
        tmp.append(i)
        tmp.append(j)
    print("q", len(l) + 1, *tmp, flush=True)
    n = II() - num
    return n


def clamp(n, smallest, largest):
    return max(smallest, min(n, largest))


def Exit():
    ans = []
    for i in range(N):
        for j in range(N):
            if B[i * N + j] >= 1:
                ans.append((i, j))
    # Time = time.perf_counter() - start
    # if Time >= 2.5:
    #     exit()
    Last_Output(ans)
    exit()


# 答えの出力
def Last_Output(l):
    tmp = []
    for i, j in l:
        tmp.append(i)
        tmp.append(j)
    print("a", len(l), *tmp, flush=True)
    n = II()
    if n == 0:
        return False
    else:
        return True


def get_Weight():
    global Weight, Bunkatu, Idx
    num = N / 3
    Idx = [(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
    Bunkatu = [[[] for i in range(3)] for j in range(3)]
    for i in range(N):
        for j in range(N):
            x = int(i // num)
            y = int(j // num)
            Bunkatu[x][y].append((i, j))
    Weight = [0] * 9
    for i in range(3):
        for j in range(3):
            Weight[i * 3 + j] = Output_Input(Bunkatu[i][j])

    avg = sum(Weight) / 9
    Weight = [i**1.5 + avg // 2 for i in Weight]


# todo 現状、クエリは独立している。->この組のクエリでパターンを絞れる、みたいな連携したクエリを送れるとよいかも
def solve():
    global B, Tarn
    B = [-100] * (N**2)

    Tarn = 0
    count = 0
    next = calc_next()

    while True:

        # 答えの候補の取得
        flag, ans_list = get_Ans(B)

        # print("# ans", len(ans_list), ans_list)
        # 絞り込めたなら
        # ans_listの数の上限の決め方を吟味する->現在期待値5だが、一回のクエリで絞り込めるならそっちがお得
        if flag and 1 <= len(ans_list) <= 5:
            for i in ans_list:
                tmp = set()
                for idx, (j, _, _, _) in enumerate(D):
                    sx, sy = i[idx]
                    for x, y in j:
                        tmp.add((x + sx, y + sy))
                Tarn += 1

                # Time = time.perf_counter() - start
                # if Time >= 2.5:
                #     exit()

                if Last_Output(tmp):
                    exit()
        # 絞り込めなかった場合
        else:
            # 次に送るクエリの候補を計算
            if flag:
                count += 1
                next = calc_next_from_ans(ans_list)
            else:
                if len(ans_list) == 0:
                    next = calc_next()
                else:
                    next = calc_next_from_ans(ans_list)

        # print("# next", len(next), next)
        if len(next) == 0:
            Exit()

        # クエリを送信、Bの更新
        if Tarn <= 150:
            num = 1
        elif Tarn <= 250:
            num = 2
        else:
            num = 3

        for i in range(num):
            while True:
                if next:
                    x, y = next.pop()
                else:
                    break

                if B[x * N + y] == -100:
                    Tarn += 1
                    B[x * N + y] = Output_Input([(x, y)])
                    break


def calc_next_from_ans(ans_list):
    next = [0] * (N**2)
    for ans in ans_list:
        tmp = set()
        for idx, (x, y) in enumerate(ans):
            for i, j in D[idx][0]:
                tmp.add((x + i, y + j))

        for i, j in tmp:
            next[i * N + j] += 1

    tmp = []
    for i in range(N):
        for j in range(N):
            if next[i * N + j] and B[i * N + j] == -100:
                tmp.append((next[i * N + j], i, j))
            elif next[i * N + j] == 0 and B[i * N + j] == -100:
                B[i * N + j] = 0
                # print(f"#c {i} {j} green")

    tmp.sort(key=lambda x: abs(x[0] - len(ans_list) / 2), reverse=True)

    return [(j, k) for i, j, k in tmp]


# 確定で0になるやつだけでなく、確定で１以上になるものなども記録できるとよい
# 送るクエリ計算
# 既に1以上確定している場所はそれを消化しなければいけないという視点


def calc_next():
    can_use = [set() for _ in range(N**2)]
    next = [0] * (N**2)
    Plus_sum = 0
    for i in range(N):
        for j in range(N):
            if B[i * N + j] >= 1:
                Plus_sum += B[i * N + j]
    count = 0
    for idx, (d, size, visited, (x_max, y_max)) in enumerate(D):

        for i in range(N - x_max):
            for j in range(N - y_max):
                if not visited[i * N + j]:
                    continue
                tmp = []
                tmp1 = []
                flag = 0
                for x, y in d:
                    if i + x in range(N) and j + y in range(N):
                        if B[(i + x) * N + (j + y)] == -100:
                            tmp.append((i + x, j + y))
                        elif B[(i + x) * N + (j + y)] >= 1:
                            tmp1.append(((i + x, j + y)))
                            flag += 1
                        else:
                            flag = -1
                            break
                    else:
                        flag = -1
                        break

                if flag >= 0 and D_size[0] - size >= Plus_sum - flag:
                    count += 1
                    for p, q in tmp:
                        next[p * N + q] += 1
                        if D_size_sum / N**2 < 0.35 and Tarn > D_size_sum * 0.35:

                            next[p * N + q] += (flag / size) ** 0.5 * 10

                    for p, q in tmp1:
                        can_use[p * N + q].add(idx)

                else:
                    visited[i * N + j] = False

    tmp = []
    # print("#", can_use)
    # print("#", B)
    for i in range(N):
        for j in range(N):
            # print("#", len(can_use[i * N + j]))
            if len(can_use[i * N + j]) == 1 and not kakutei[i * N + j]:

                idx = can_use[i * N + j].pop()

                kakutei[i * N + j] = True
                d, size, visited, (x_max, y_max) = D[idx]
                # print("#", i, j, idx, d)
                for x in range(N - x_max):
                    for y in range(N - y_max):
                        if not visited[x * N + y]:
                            continue
                        flag = 0
                        for p, q in d:
                            if x + p in range(N) and y + q in range(N):
                                if (
                                    B[(x + p) * N + (y + q)] == -100
                                    or B[(x + p) * N + (y + q)] >= 1
                                ):
                                    if x + p == i and y + q == j:
                                        flag = 1
                                else:
                                    flag = -1
                                    break
                            else:
                                flag = -1
                                break
                        if flag <= 0:

                            visited[x * N + y] = False
                        # else:
                        #     print("#", x, y)

            if next[i * N + j] and B[i * N + j] == -100:
                tmp.append((next[i * N + j], i, j))
            elif next[i * N + j] == 0 and B[i * N + j] == -100:
                B[i * N + j] = 0
                # print(f"#c {i} {j} blue")
    if D_size_sum / N**2 < 0.35 and Tarn > D_size_sum * 0.35:
        print("# another")
        tmp.sort(key=lambda x: x[0])
    else:
        tmp.sort(key=lambda x: abs(x[0] - count / 2), reverse=True)

    return [(j, k) for _, j, k in tmp]


# 答えの案の出力
# 候補数が多すぎる場合打ち切っているが、上手いこと計算すれば、打ち切らずに済場合がある？
# 計算結果を保存しておくことで、同じ計算を防いで高速化＆効率化できるかも
# 既に確定した部分を利用して、多めにクエリを送って引き算すればコストを抑えられる（誤差が生まれるので、その誤算に注意する必要あり）


def get_Ans(B):
    deq = deque()
    B_n = B[:]
    deq.append((0, B_n, []))
    ans = []
    while deq:
        idx, B, prv = deq.popleft()
        shape = D[idx][0]
        visited = D[idx][2]
        x_max, y_max = D[idx][3]
        if idx == 1:
            big_flag = False
        else:
            big_flag = True
        Sum = 0
        for i in range(N):
            for j in range(N):
                if B[i * N + j] >= 1:
                    Sum += 1

        for i in range(N - x_max):
            for j in range(N - y_max):
                if not visited[i * N + j]:
                    continue
                flag = True
                for x, y in shape:
                    if (
                        i + x in range(N)
                        and j + y in range(N)
                        and (
                            B[(i + x) * N + (j + y)] <= -100
                            or B[(i + x) * N + (j + y)] >= 1
                        )
                    ):
                        pass
                    else:
                        flag = False
                        break

                if flag:
                    prv_n = prv[:]
                    prv_n.append((i, j))
                    B_n = B[:]
                    if idx == M - 1:
                        for x, y in shape:
                            B_n[(i + x) * N + (j + y)] -= 1
                        if max(B_n) <= 0:
                            big_flag = True
                            ans.append(prv_n)
                        if len(ans) >= 3 * 10**5:
                            return False, ans
                    else:
                        tmp = 0
                        for x, y in shape:
                            B_n[(i + x) * N + (j + y)] -= 1
                            if B_n[(i + x) * N + (j + y)] >= 0:
                                tmp += 1

                        if max(B_n) <= M - idx - 1 and D_size[idx + 1] >= Sum - tmp:
                            big_flag = True
                            deq.append((idx + 1, B_n, prv_n))

                        if len(deq) >= limit:
                            return False, []
        if big_flag == False and D[0][2][x * N + y]:
            x, y = prv[0]
            D[0][2][x * N + y] = False
            # print(f"#c {x} {y} red")

    return True, ans


# import random

# import time


def main():
    # global start
    # start = time.perf_counter()
    FirstInput()
    solve()


if __name__ == "__main__":
    main()

File: ahc030/a/tools/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
)

var sc = bufio.NewScanner(os.Stdin)
var wr = bufio.NewWriter(os.Stdout)

func main() {
	defer wr.Flush()
	sc.Split(bufio.ScanWords)
	sc.Buffer([]byte{}, math.MaxInt32)
	// out("# start")
	FirstInput()
	solve()
	// use getI(), getS(), getInts(), getF()
}

var N int
var M int
var E float64
var D_zahyou [][][]int
var D_len_zahyou []int
var D_visited [][]bool
var D_maxmin [][]int
var kakutei []bool
var limit int
var D_size_sum int
var N_2 int
var D_size []int
var Tarn int

func FirstInput() {
	N, M, E = getI(), getI(), getF()
	N_2 = N * N

	kakutei = make([]bool, N_2)
	D_size_sum = 0
	limit = 3000000 / N_2
	D_zahyou = make([][][]int, 0, M)
	D_len_zahyou = make([]int, 0, M)
	D_visited = make([][]bool, 0, M)
	D_maxmin = make([][]int, 0, M)
	for i := 0; i < M; i++ {
		num := getI()
		tmp := getInts(num * 2)
		x_max, y_max := 0, 0
		zahyou := make([][]int, 0, num)
		for j := 1; j < num*2; j += 2 {
			x_max = max(x_max, tmp[j-1])
			y_max = max(y_max, tmp[j])
			zahyou = append(zahyou, []int{tmp[j-1], tmp[j]})
		}
		D_zahyou = append(D_zahyou, zahyou)
		D_len_zahyou = append(D_len_zahyou, len(zahyou))
		D_visited = append(D_visited, make([]bool, N_2))
		D_maxmin = append(D_maxmin, []int{x_max, y_max})
		D_size_sum += len(zahyou)
	}
	sort.Slice(D_zahyou, func(i, j int) bool {
		return D_len_zahyou[i] > D_len_zahyou[j]
	})
	sort.Slice(D_len_zahyou, func(i, j int) bool {
		return D_len_zahyou[i] > D_len_zahyou[j]
	})
	sort.Slice(D_visited, func(i, j int) bool {
		return D_len_zahyou[i] > D_len_zahyou[j]
	})
	sort.Slice(D_maxmin, func(i, j int) bool {
		return D_len_zahyou[i] > D_len_zahyou[j]
	})
	D_size = make([]int, M)
	copy(D_size, D_len_zahyou)

	// fmt.Println("#", D_len_zahyou, D_maxmin, D_len_zahyou)

	for i := M - 2; i >= 0; i-- {
		D_size[i] += D_size[i+1]
	}

	// fmt.Println("#", D_size)
}

func OutputInput(l [][]int) int {
	tmp := make([]int, 0, D_size_sum)
	for _, pair := range l {
		tmp = append(tmp, pair[0])
		tmp = append(tmp, pair[1])
	}
	fmt.Printf("q %d", len(l))
	for _, val := range tmp {
		fmt.Printf(" %d", val)
	}
	fmt.Println()

	n := getI()
	return n
}

func Exit(B []int) {
	ans := make([][2]int, 0, D_size_sum)
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			if B[i*N+j] >= 1 {
				ans = append(ans, [2]int{i, j})
			}
		}
	}

	LastOutput(ans)
	os.Exit(0)
}

func LastOutput(l [][2]int) bool {
	tmp := make([]int, 0, D_size_sum*2+10)
	for _, pair := range l {
		tmp = append(tmp, pair[0])
		tmp = append(tmp, pair[1])
	}
	fmt.Printf("a %d", len(l))
	for _, val := range tmp {
		fmt.Printf(" %d", val)
	}
	fmt.Println()
	n := getI()
	if n == 0 {
		return false
	} else {
		return true
	}
}

func solve() {
	B := make([]int, N_2)
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			B[i*N+j] = -100
		}
	}
	Tarn = 0
	count := 0
	next := calcNext(B)
	// fmt.Println("#", next)
	for {
		// 答えの候補の取得
		flag, ansList := getAns(B)
		fmt.Println("#", flag, len(ansList))

		// 絞り込めたなら
		// ansListの数の上限の決め方を吟味する->現在期待値5だが、一回のクエリで絞り込めるならそっちがお得
		if flag && 1 <= len(ansList) && len(ansList) <= 5 {
			for _, i := range ansList {
				tmp := make(map[[2]int]struct{})
				for idx, j := range D_zahyou {
					sx, sy := i[idx][0], i[idx][1]
					for _, xy := range j {
						tmp[[2]int{xy[0] + sx, +xy[1] + sy}] = struct{}{}
					}
				}
				Tarn++

				tmp1 := make([][2]int, 0, len(tmp))
				for key, _ := range tmp {
					tmp1 = append(tmp1, key)
				}
				if LastOutput(tmp1) {
					os.Exit(0)
				}
			}
		} else {
			// 絞り込めなかった場合
			// 次に送るクエリの候補を計算
			if flag {
				count++
				next = calcNextFromAns(ansList, B)
			} else {
				if len(ansList) == 0 {
					next = calcNext(B)
				} else {
					next = calcNextFromAns(ansList, B)
				}
			}
		}

		if len(next) == 0 {
			Exit(B)
		}

		// クエリを送信、Bの更新
		num := 1
		if Tarn > 150 {
			num = 2
		}
		if Tarn > 250 {
			num = 3
		}

		for i := 0; i < num; i++ {
			for {
				if len(next) == 0 {
					break
				}
				x, y := next[0][0], next[0][1]
				next = next[1:]

				if B[x*N+y] == -100 {
					Tarn++
					B[x*N+y] = OutputInput([][]int{{x, y}})
					break
				}
			}
		}
	}
}

func calcNextFromAns(ansList [][][2]int, B []int) [][2]int {
	next := make([]int, N*N)
	for _, ans := range ansList {
		tmp := make(map[[2]int]struct{})
		for idx, xy := range ans {
			// out(xy)
			x, y := xy[0], xy[1]
			for _, ij := range D_zahyou[idx] {
				tmp[[2]int{ij[0] + x, ij[1] + y}] = struct{}{}
			}
		}

		for ij, _ := range tmp {
			i, j := ij[0], ij[1]
			next[i*N+j] += 1
		}
	}

	tmp := make([][3]int, 0, N_2)
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			if next[i*N+j] > 0 && B[i*N+j] == -100 {
				tmp = append(tmp, [3]int{next[i*N+j], i, j})
			} else if next[i*N+j] == 0 && B[i*N+j] == -100 {
				B[i*N+j] = 0
			}
		}
	}

	sort.Slice(tmp, func(i, j int) bool {
		return abs(tmp[i][0]-len(ansList)/2) > abs(tmp[j][0]-len(ansList)/2)
	})

	tmp1 := make([][2]int, 0, N_2)
	for _, i := range tmp {
		j, k := i[1], i[2]
		tmp1 = append(tmp1, [2]int{j, k})
	}

	return tmp1
}

func calcNext(B []int) [][2]int {
	canUse := make([]map[int]bool, N*N)
	for i := 0; i < N_2; i++ {
		canUse[i] = make(map[int]bool)
	}
	next := make([]float64, N*N)
	plusSum := 0

	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			if B[i*N+j] >= 1 {
				plusSum += B[i*N+j]
			}
		}
	}

	count := 0
	for idx, d := range D_zahyou {
		size := D_len_zahyou[idx]
		visited := D_visited[idx]
		xMax, yMax := D_maxmin[idx][0], D_maxmin[idx][1]
		for i := 0; i < N-xMax; i++ {
			for j := 0; j < N-yMax; j++ {
				if visited[i*N+j] {
					continue
				}

				tmp := make([][2]int, 0, N_2)
				tmp1 := make([][2]int, 0, N_2)
				flag := 0

				for _, coord := range d {
					x, y := coord[0], coord[1]
					if i+x >= 0 && i+x < N && j+y >= 0 && j+y < N {
						if B[(i+x)*N+(j+y)] == -100 {
							tmp = append(tmp, [2]int{(i + x), (j + y)})
						} else if B[(i+x)*N+(j+y)] >= 1 {
							tmp1 = append(tmp1, [2]int{(i + x), (j + y)})
							flag++
						} else {
							flag = -1
							break
						}
					} else {
						flag = -1
						break
					}
				}

				if flag >= 0 && D_size[0]-size >= plusSum-flag {
					count++
					for _, coord := range tmp {
						p, q := coord[0], coord[1]
						next[p*N+q]++
						if float64(D_size_sum)/float64(N_2) < 0.35 && float64(Tarn) > float64(D_size_sum)*0.35 {
							next[p*N+q] += math.Pow(1-float64(flag)/float64(size), 0.5) * 10
						}
					}

					for _, coord := range tmp1 {
						p, q := coord[0], coord[1]
						canUse[p*N+q][idx] = true
					}
				} else {
					visited[i*N+j] = true
				}
			}
		}
	}

	tmp := make([][3]int, 0, N_2)

	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			if len(canUse[i*N+j]) == 1 && !kakutei[i*N+j] {
				var idx int
				for key := range canUse[i*N+j] {
					idx = key
				}
				kakutei[i*N+j] = true
				d, visited, xMax, yMax := D_zahyou[idx], D_visited[idx], D_maxmin[idx][0], D_maxmin[idx][1]

				for x := 0; x < N-xMax; x++ {
					for y := 0; y < N-yMax; y++ {
						if visited[x*N+y] {
							continue
						}
						flag := 0
						for _, pq := range d {
							p, q := pq[0], pq[1]
							if x+p >= 0 && x+p < N && y+q >= 0 && y+q < N {
								if B[(x+p)*N+(y+q)] == -100 || B[(x+p)*N+(y+q)] >= 1 {
									if x+p == i && y+q == j {
										flag = 1
									}
								} else {
									flag = -1
									break
								}
							} else {
								flag = -1
								break
							}
						}
						if flag <= 0 {
							visited[x*N+y] = true
						}
					}
				}
			}

			if next[i*N+j] != 0 && B[i*N+j] == -100 {
				tmp = append(tmp, [3]int{int(next[i*N+j]), i, j})
			} else if next[i*N+j] == 0 && B[i*N+j] == -100 {
				B[i*N+j] = 0
				// fmt.Printf("#c %d %d blue\n", i, j)
			}
			// fmt.Println("#", tmp)
		}
	}

	if float64(D_size_sum)/float64(N_2) < 0.35 && float64(Tarn) > float64(D_size_sum)*0.35 {
		sort.Slice(tmp, func(i, j int) bool {
			return tmp[i][0] > tmp[j][0]
		})
	} else {
		sort.Slice(tmp, func(i, j int) bool {
			return abs(tmp[i][0]-count/2) < abs(tmp[j][0]-count/2)
		})
	}

	result := make([][2]int, 0, N_2)
	for _, t := range tmp {

		i, j := t[1], t[2]
		result = append(result, [2]int{i, j})
	}
	return result
}

func getAns(B []int) (bool, [][][2]int) {
	var deq []struct {
		idx int
		B   []int
		prv [][2]int
	}
	B_n := make([]int, len(B))
	copy(B_n, B)
	deq = append(deq, struct {
		idx int
		B   []int
		prv [][2]int
	}{0, B_n, [][2]int{}})

	ans := make([][][2]int, 0, 300000)

	for len(deq) > 0 {

		front := deq[0]
		deq = deq[1:]
		idx, B, prv := front.idx, front.B, front.prv
		shape := D_zahyou[idx]
		visited := D_visited[idx]
		xMax, yMax := D_maxmin[idx][0], D_maxmin[idx][1]
		Sum := 0
		for i := 0; i < N; i++ {
			for j := 0; j < N; j++ {
				if B[i*N+j] >= 1 {
					Sum++
				}
			}
		}
		for i := 0; i < N-xMax; i++ {
			for j := 0; j < N-yMax; j++ {
				if visited[i*N+j] {
					continue
				}
				flag := true
				for _, coord := range shape {
					x, y := coord[0], coord[1]
					if i+x >= 0 && i+x < N && j+y >= 0 && j+y < N && (B[(i+x)*N+(j+y)] <= -100 || B[(i+x)*N+(j+y)] >= 1) {
					} else {
						flag = false
						break
					}
				}
				if flag {
					prv_n := make([][2]int, len(prv))
					copy(prv_n, prv)
					prv_n = append(prv_n, [2]int{i, j})
					B_n := make([]int, len(B))
					copy(B_n, B)
					if idx == M-1 {
						for _, coord := range shape {
							x, y := coord[0], coord[1]
							B_n[(i+x)*N+(j+y)]--
						}
						if max(B_n...) <= 0 {
							ans = append(ans, prv_n)
						}
						if len(ans) >= 300000 {
							return false, ans
						}
					} else {
						tmp := 0
						for _, coord := range shape {
							x, y := coord[0], coord[1]
							B_n[(i+x)*N+(j+y)]--
							if B_n[(i+x)*N+(j+y)] >= 0 {
								tmp++
							}
						}
						if max(B_n...) <= M-idx-1 && D_size[idx+1] >= Sum-tmp {
							deq = append(deq, struct {
								idx int
								B   []int
								prv [][2]int
							}{idx + 1, B_n, prv_n})
						}
						if len(deq) >= limit {
							return false, nil
						}
					}
				}
			}
		}
	}
	return true, ans
}

func out(x ...interface{}) {
	fmt.Fprintln(wr, x...)
}

func getI() int {
	sc.Scan()
	i, e := strconv.Atoi(sc.Text())
	if e != nil {
		panic(e)
	}
	return i
}

func getF() float64 {
	sc.Scan()
	i, e := strconv.ParseFloat(sc.Text(), 64)
	if e != nil {
		panic(e)
	}
	return i
}

func getInts(N int) []int {
	ret := make([]int, N)
	for i := 0; i < N; i++ {
		ret[i] = getI()
	}
	return ret
}

func getS() string {
	sc.Scan()
	return sc.Text()
}

// 型変換
func toStr(num int) string {
	return strconv.Itoa(num)
}

func toInt(s string) int {
	i, _ := strconv.Atoi(s)
	return i
}

// 重複削除
func toSet(arr []int) []int {
	m := make(map[int]struct{})
	for _, v := range arr {
		m[v] = struct{}{}
	}
	var newArr []int
	for k, _ := range m {
		newArr = append(newArr, k)
	}
	return newArr
}

func toMap(slice []int) map[int]bool {
	m := make(map[int]bool)
	for _, s := range slice {
		m[s] = true
	}
	return m
}

func inSlice(slice []int, key int) bool {
	for _, i := range slice {
		if i == key {
			return true
		}
	}
	return false
}

func inMap(Map map[int]bool, key int) bool {
	_, exist := Map[key]
	return exist
}

func sum(a ...int) int {
	temp := 0
	for _, v := range a {
		temp += v
	}
	return temp
}

func max(a ...int) int {
	temp := a[0]
	for _, v := range a {
		if temp < v {
			temp = v
		}
	}
	return temp
}

func min(a ...int) int {
	temp := a[0]
	for _, v := range a {
		if temp > v {
			temp = v
		}
	}
	return temp
}

func chmin(a *int, b int) bool {
	if *a < b {
		return false
	}
	*a = b
	return true
}

func chmax(a *int, b int) bool {
	if *a > b {
		return false
	}
	*a = b
	return true
}

func asub(a, b int) int {
	if a > b {
		return a - b
	}
	return b - a
}

func abs(a int) int {
	if a >= 0 {
		return a
	}
	return -a
}

func lowerBound(a []int, x int) int {
	idx := sort.Search(len(a), func(i int) bool {
		return a[i] >= x
	})
	return idx
}

func upperBound(a []int, x int) int {
	idx := sort.Search(len(a), func(i int) bool {
		return a[i] > x
	})
	return idx
}
